src/calculus/sum.jl

```diff
@@ -0,0 +1,59 @@
+export Sum
+
+"""
+**Sum of functions**
+
+    Sum(f₁,…,fₖ)
+
+Given functions `f₁` to `fₖ`, returns their sum
+
+```math
+g(x) = ∑_\{i=1\}^k f_i(x).
+```
+
+The functions `fᵢ` should all have the same domain, which will be the domain of
+the result. If all `fᵢ` have gradients then this function will also have a
+gradient.
+
+"""
+
+immutable Sum{T <: Tuple} <: ProximableFunction fs::T end
+
+Sum(fs::Vararg{ProximableFunction}) = Sum((fs...))
+
+# note: is_prox_accurate false because prox in general doesn't exist?
+is_prox_accurate(f::Sum) = false
+is_convex(f::Sum) = all(is_convex.(f.fs))
+is_set(f::Sum) = all(is_set.(f.fs))
+is_cone(f::Sum) = all(is_cone.(f.fs))
+is_affine(f::Sum) = all(is_affine.(f.fs))
+is_smooth(f::Sum) = all(is_smooth.(f.fs))
+is_quadratic(f::Sum) = all(is_quadratic.(f.fs))
+is_generalized_quadratic(f::Sum) = all(is_generalized_quadratic.(f.fs))
+is_strongly_convex(f::Sum) = all(is_convex.(f.fs)) && any(is_strongly_convex.(f.fs))
+
+function (sumobj::Sum)(x)
+    sum = 0.0
```

Comment:
To answer your question, here you could be doing 
```
function (sumobj::Sum)(x::AbstractArray{T}) where {R, T <: Union{R, Complex{R}}}
    sum = zero(R)
    ...
```

And similarly in `gradient!`.

This should be type stable provided that all component functions evaluate satisfy the same assumption, i.e. if the evaluate to R where R is the basic real type of their argument. I’m not sure that *all* the functions implemented satisfy this, but if they don’t then an issue should be opened and this should be fixed.

Reply:
This looks a bit weird, if `T=Complex{Float64}`, then can't `R = Complex{Float64} <: Union{Complex{Float64}, Complex{Complex{Float64}}}`? I'm not sure how `T` gets decided here, Maybe we should at least say `R<:Real`?

Reply:
Yes, of course :-)